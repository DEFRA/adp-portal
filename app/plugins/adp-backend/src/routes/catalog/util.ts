export function createTransformerTitle(title: string, alias?: string) {
  const titleValue = alias ? `${title} ` + `(${alias})` : title;
  return titleValue;
}
export const ARMS_LENGTH_BODY_ID_ANNOTATION =
  'adp.defra.gov.uk/arms-length-body-id';
export const DELIVERY_PROGRAMME_ID_ANNOTATION =
  'adp.defra.gov.uk/delivery-programme-id';
export const DELIVERY_PROJECT_ID_ANNOTATION =
  'adp.defra.gov.uk/delivery-project-id';
