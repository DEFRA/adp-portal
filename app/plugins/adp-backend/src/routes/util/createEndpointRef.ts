import {
  type ServiceRef,
  createServiceFactory,
  createServiceRef,
} from '@backstage/backend-plugin-api';
import type { Request, RequestHandler } from 'express';
import { randomUUID } from 'node:crypto';
import type { HandlerResult } from './HandlerResult';
import type { ServiceRefsToInstances } from './ServiceRefsToInstances';
import { routerResultsRef, type RouterResults } from './routerResultsRef';

type Endpoint<
  P,
  ResBody,
  ReqBody,
  ReqQuery,
  Locals extends Record<string, any>,
> = (
  request: Request<P, ResBody, ReqBody, ReqQuery, Locals>,
) => HandlerResult | PromiseLike<HandlerResult>;

interface EndpointOptions<
  Dependencies extends Record<string, ServiceRef<unknown>>,
  P,
  ResBody,
  ReqBody,
  ReqQuery,
  Locals extends Record<string, any>,
> {
  deps: Dependencies;
  factory(options: {
    deps: ServiceRefsToInstances<Dependencies>;
    responses: RouterResults;
  }):
    | Endpoint<P, ResBody, ReqBody, ReqQuery, Locals>
    | PromiseLike<Endpoint<P, ResBody, ReqBody, ReqQuery, Locals>>;
}

export function createEndpointRef<
  P,
  ResBody,
  ReqBody,
  ReqQuery,
  Locals extends Record<string, any>,
  Dependencies extends Record<string, ServiceRef<unknown>>,
>(
  options: EndpointOptions<Dependencies, P, ResBody, ReqBody, ReqQuery, Locals>,
) {
  return createServiceRef<
    RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>
  >({
    id: `adp.router.endpoint.${randomUUID()}`,
    scope: 'plugin',
    defaultFactory(service) {
      const routerResultsKey = randomUUID() as 'routerResults';
      return Promise.resolve(
        createServiceFactory<
          RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>,
          RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>,
          any
        >({
          service,
          deps: { ...options.deps, [routerResultsKey]: routerResultsRef },
          async factory({ [routerResultsKey]: responses, ...deps }) {
            const handler = await options.factory({
              deps: deps as Dependencies,
              responses,
            });
            return voidify(async (req, res, next) => {
              try {
                const result = await handler(req);
                await result.writeTo(res);
              } catch (err) {
                next(err);
              }
            });
          },
        }),
      );
    },
  });
}

function voidify<This, Args extends unknown[]>(
  fn: (this: This, ...args: Args) => unknown,
): (this: This, ...args: Args) => void {
  return function voided(...args) {
    void fn.call(this, ...args);
  };
}
