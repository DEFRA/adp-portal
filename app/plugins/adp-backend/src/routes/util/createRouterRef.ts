import {
  type ServiceRef,
  createServiceFactory,
  createServiceRef,
} from '@backstage/backend-plugin-api';
import type { Router } from 'express';
import { randomUUID } from 'node:crypto';
import { routerFactoryRef } from './routerFactoryRef';
import { type RouterResults, routerResultsRef } from './routerResultsRef';
import type { ServiceRefsToInstances } from './ServiceRefsToInstances';

interface RouterFactoryOptions<
  Dependencies extends Record<string, ServiceRef<unknown, 'plugin' | 'root'>>,
> {
  deps: Dependencies;
  factory(options: {
    router: Router;
    deps: ServiceRefsToInstances<Dependencies>;
    responses: RouterResults;
  }): void | Promise<void>;
}

export function createRouterRef<
  Dependencies extends Record<string, ServiceRef<unknown, 'plugin' | 'root'>>,
>(options: RouterFactoryOptions<Dependencies>) {
  return createServiceRef<Router>({
    id: `adp.router.${randomUUID()}`,
    scope: 'plugin',
    defaultFactory(service) {
      return Promise.resolve(createRouterFactory(service, options));
    },
  });
}
function createRouterFactory<
  Dependencies extends Record<string, ServiceRef<unknown, 'plugin' | 'root'>>,
>(
  service: ServiceRef<Router, 'plugin'>,
  options: RouterFactoryOptions<Dependencies>,
) {
  const routerFactoryKey = randomUUID() as 'routerFactory';
  const routerResultsKey = randomUUID() as 'routerResults';
  return createServiceFactory<Router, Router, any>({
    service,
    deps: {
      ...options.deps,
      [routerFactoryKey]: routerFactoryRef,
      [routerResultsKey]: routerResultsRef,
    },
    async factory({
      [routerFactoryKey]: routerFactory,
      [routerResultsKey]: routerResults,
      ...deps
    }) {
      const router = routerFactory();
      await options.factory({
        router,
        deps: deps as Dependencies,
        responses: routerResults,
      });
      return router;
    },
  });
}
