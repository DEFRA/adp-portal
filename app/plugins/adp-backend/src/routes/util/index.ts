export * from './createEndpointRef';
export * from './createRouterRef';
export * from './HandlerResult';
export * from './routerFactoryRef';
export * from './routerResultsRef';
