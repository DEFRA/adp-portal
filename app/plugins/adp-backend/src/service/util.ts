import { Response } from 'express';
import { ValidationErrorMapping } from '@internal/plugin-adp-common';
import { InputError } from '@backstage/errors';
import { z } from 'zod';

export function respond<Request, Success, Error extends string>(
  request: Request,
  response: Response,
  message:
    | { success: true; value: Success }
    | { success: false; errors: Error[] },
  mapping: ValidationErrorMapping<Request, Error>,
  statuses?: { ok?: number; badRequest?: number },
) {
  if (!message.success) {
    response.status(statuses?.badRequest ?? 400).json({
      errors: message.errors.map(e => mapping[e](request)),
    });
  } else if (message.value === undefined) {
    response.status(statuses?.ok ?? 204).end();
  } else {
    response.status(statuses?.ok ?? 200).json(message.value);
  }
}

export function createParser<T>(schema: z.ZodType<T>) {
  return (body: unknown) => {
    const result = schema.safeParse(body);
    if (result.success) return result.data;
    throw new InputError(result.error.message);
  };
}

export type SafeResult<Success, Error extends PropertyKey> =
  | { success: true; value: Success }
  | { success: false; errors: Array<Error | 'unknown'> };

export const emptyUUID = '00000000-0000-0000-0000-000000000000';
export type UUID = `${string}-${string}-${string}-${string}-${string}`;

export function isUUID(value: string): value is UUID {
  return /^[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12}$/i.test(
    value,
  );
}

export async function checkMany<
  T extends Record<string, boolean | Promise<boolean>>,
>(checks: T): Promise<SafeResult<undefined, keyof T>> {
  const results: Array<[keyof T, boolean]> = await Promise.all(
    Object.entries(checks).map(async e => [e[0], await e[1]] as const),
  );
  const failed = results.filter(x => x[1]);
  if (failed.length === 0) return { success: true, value: undefined };

  return { success: false, errors: failed.map(x => x[0]) };
}

export function containsAnyValue(obj: object) {
  return Object.entries(obj).some(e => e[1] !== undefined);
}
