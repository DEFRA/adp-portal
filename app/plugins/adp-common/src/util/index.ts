export * from './annotations';
export * from './createName';
export * from './normalizeUsername';
