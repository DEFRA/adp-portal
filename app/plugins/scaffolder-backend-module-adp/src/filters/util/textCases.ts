interface TextCase {
  toWords(input: string): string[];
  fromWords(words: string[]): string;
}

function splitBy(separator: string): TextCase {
  return {
    toWords(input) {
      return input.split(separator).map(w => w.toLowerCase());
    },
    fromWords(words) {
      return words.map(w => w.toLowerCase()).join(separator);
    },
  };
}

const core = {
  pascal: {
    toWords(input) {
      return input.split(/(?=[A-Z])/g).map(w => w.toLowerCase());
    },
    fromWords(words) {
      return words
        .map(w => w[0].toUpperCase() + w.slice(1).toLowerCase())
        .join('');
    },
  },
  sentence: splitBy(' '),
  kebab: splitBy('-'),
  snake: splitBy('_'),
  dot: splitBy('.'),
} satisfies Record<string, TextCase>;

export const textCases = {
  ...core,
  camel: {
    toWords(input) {
      return core.pascal.toWords(input);
    },
    fromWords(words) {
      const result = core.pascal.fromWords(words);
      return result[0].toLowerCase() + result.slice(1);
    },
  },
} as const satisfies Record<string, TextCase>;
