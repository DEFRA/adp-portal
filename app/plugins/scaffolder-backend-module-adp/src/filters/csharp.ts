import type { JsonValue } from '@backstage/types';
import { csharpConverters } from './util/csharpConverters';

const converterNames = Object.keys(
  csharpConverters,
) as readonly (keyof typeof csharpConverters)[];

export function csharp(input: JsonValue, kind: JsonValue = 'value') {
  const kindStr = String(kind);
  if (!isConverter(kindStr)) throw new Error(`Unsupported kind ${kindStr}`);
  return csharpConverters[kindStr](input);
}

function isConverter(value: string): value is keyof typeof csharpConverters {
  return converterNames.includes(value as any);
}
