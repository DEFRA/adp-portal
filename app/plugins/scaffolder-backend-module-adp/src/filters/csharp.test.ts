import type { JsonValue } from '@backstage/types';
import { csharpConverters } from './util/csharpConverters';
import { csharp } from './csharp';

const converters = Object.keys(
  csharpConverters,
) as readonly (keyof typeof csharpConverters)[];
const inputs: JsonValue[] = [
  'My Cool Project',
  'My-Cool-Project',
  'my_cool_project',
  'test@email.com',
  '123 xyz',
  true,
  false,
  123.456,
  { some: { cool: 'object' } },
  null,
];

interface TestCase {
  readonly input: JsonValue;
  readonly type?: JsonValue;
  readonly inputJSON: string;
}
describe('csharp', () => {
  it.each<TestCase>(
    inputs.flatMap(input =>
      converters.map(type => ({
        input,
        type,
        inputJSON: JSON.stringify(input),
      })),
    ),
  )('Should format $inputJSON as $type', ({ input, type }) => {
    // act
    const actual = csharp(input, type);

    // assert
    expect(actual).toMatchSnapshot();
  });
});
