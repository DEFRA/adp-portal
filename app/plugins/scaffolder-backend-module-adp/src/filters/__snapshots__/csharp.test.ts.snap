// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp Should format "123 xyz" as identifier 1`] = `"_123_xyz"`;

exports[`csharp Should format "123 xyz" as namespace 1`] = `"_123.Xyz"`;

exports[`csharp Should format "123 xyz" as value 1`] = `""123 xyz""`;

exports[`csharp Should format "My Cool Project" as identifier 1`] = `"My_Cool_Project"`;

exports[`csharp Should format "My Cool Project" as namespace 1`] = `"My.Cool.Project"`;

exports[`csharp Should format "My Cool Project" as value 1`] = `""My Cool Project""`;

exports[`csharp Should format "My-Cool-Project" as identifier 1`] = `"My_Cool_Project"`;

exports[`csharp Should format "My-Cool-Project" as namespace 1`] = `"My.Cool.Project"`;

exports[`csharp Should format "My-Cool-Project" as value 1`] = `""My-Cool-Project""`;

exports[`csharp Should format "my_cool_project" as identifier 1`] = `"My_cool_project"`;

exports[`csharp Should format "my_cool_project" as namespace 1`] = `"My_cool_project"`;

exports[`csharp Should format "my_cool_project" as value 1`] = `""my_cool_project""`;

exports[`csharp Should format "test@email.com" as identifier 1`] = `"Test_email_com"`;

exports[`csharp Should format "test@email.com" as namespace 1`] = `"Test_email.Com"`;

exports[`csharp Should format "test@email.com" as value 1`] = `""test@email.com""`;

exports[`csharp Should format {"some":{"cool":"object"}} as identifier 1`] = `"__object_Object_"`;

exports[`csharp Should format {"some":{"cool":"object"}} as namespace 1`] = `"__object.Object_"`;

exports[`csharp Should format {"some":{"cool":"object"}} as value 1`] = `
"new 
{
  some = new 
{
  cool = "object"
}
}"
`;

exports[`csharp Should format 123.456 as identifier 1`] = `"_123_456"`;

exports[`csharp Should format 123.456 as namespace 1`] = `"_123._456"`;

exports[`csharp Should format 123.456 as value 1`] = `"123.456"`;

exports[`csharp Should format false as identifier 1`] = `"False"`;

exports[`csharp Should format false as namespace 1`] = `"False"`;

exports[`csharp Should format false as value 1`] = `"false"`;

exports[`csharp Should format null as identifier 1`] = `"Null"`;

exports[`csharp Should format null as namespace 1`] = `"Null"`;

exports[`csharp Should format null as value 1`] = `"null"`;

exports[`csharp Should format true as identifier 1`] = `"True"`;

exports[`csharp Should format true as namespace 1`] = `"True"`;

exports[`csharp Should format true as value 1`] = `"true"`;
